cli/upload.py

```diff
@@ -83,65 +84,154 @@ def _open_file_mapping(extra_metadata: dict) -> Dict[str, BinaryIO]:
+significantly faster if you enable parallel composite uploads. This
+feature can be enabled by editing the
+"parallel_composite_upload_threshold" value in your .boto
+configuration file. However, note that if you do this large files will
+be uploaded as `composite objects
+<https://cloud.google.com/storage/docs/composite-objects>`_,which
+means that any user who downloads such objects will need to have a
+compiled crcmod installed (see "gsutil help crcmod"). This is because
+without a compiled crcmod, computing checksums on composite objects is
+so slow that gsutil disables downloads of composite objects.""".split(
+            "\n"
+        ),
+    )
+)
+
+
+def _start_procs(upload_info: api.UploadInfo, xlsx: str) -> list:
+    """
+    Starts multiple "gsutil cp" subprocesses.
+    Returns a list of subprocess.Popen objects
+    """
+    procs = []
+
+    for src, dst in _compose_file_mapping(upload_info, xlsx):
+
+        # Construct the upload command
+        gsutil_args = ["gsutil", "-m", "cp", src, dst]
+
+        try:
+            # Run the upload command
+            procs.append(
+                subprocess.Popen(
+                    gsutil_args,
+                    universal_newlines=True,
+                    bufsize=1,  # line buffered so we can read output line by line
+                    stdout=subprocess.DEVNULL,
+                    stderr=subprocess.PIPE,
+                )
+            )
+
+        except OSError as e:
+
+            # stopping already created processes
+            for p in procs:
+                p.kill()
+
+            _handle_upload_exc(e)
+    return procs
+
+
+def _wait_for_upload(procs: list) -> bool:
+    """
+    Waits for all subprocesses and prints their stderr streams.
+    Returns bool - if an error has occurred during any if uploads
+    """
+
+    finished = set()
+    errored = False
+    while len(finished) != len(procs) and not errored:
+        for i, p in enumerate(procs):
+
+            if p not in finished and p.poll() != None:
+                finished.add(i)
+
+                if p.returncode != 0:
+                    errored = True
+
+            errline = p.stderr.readline()
+
+            # skipping "large file" warnings
+            if errline.strip() in _IGNORED_WARN_LINES:
+                continue
+
+            if errline and len(errline) > 2:  # skipping '* ' spinner lines
+
+                # changing output from always "[0/1 files]" - due to manual
+                # parallelization of gsutil to "[x/y files]"
+                if errline.split("]", 1)[0].endswith("/1 files"):
+                    errline = errline.split("]", 1)[1]
+                print(f"[{len(finished)}/{len(procs)} files] {i+1}: " + errline, end="")
+
+    print(f"[{len(finished)}/{len(procs)} files]")
+    return errored
+
+
 def _gsutil_assay_upload(upload_info: api.UploadInfo, xlsx: str):
     """
     Upload local assay data to GCS using gsutil.
     """
-    workspace = _get_workspace_path()
```

Comment:
If the `workspace` directory is no longer used by the CLI, should we remove this concept from the CLI code entirely? E.g., here: https://github.com/CIMAC-CIDC/cidc-cli/blob/c2669fda3383e9c8978645ff6b7ad47a4bef0cb4/cli/config.py#L11

Reply:
sure, we don't need it (for now at least - might be useful if we decide to work on resumable uploads though)

Reply:
Ok, no harm in leaving it in.